### PR TITLE
Move Snowflake queries onto the dbt AdapterBackedSqlClient

### PR DESCRIPTION
**Pull request by tlento**
It's time to enable Snowflake queries for our first class integration
with dbt projects!

This PR generalizes the initialization of the AdapterBackedSqlClient
to allow for easy expansion to other supported adapter types, and
demonstrates the capability by enabling support for Snowflake.

Due to Snowflake's insistence on storing database objects with fully
uppercase identifiers by default, and making queries against them
case sensitive..... sometimes? Maybe? .... we need to update a lot
of test assertion checks and a few of our accessors.

The only production impact is on the list_dimensions method, which
renders a column in a result dataframe that it selects by name. In
the longer term we likely need a better story around identifier
escaping for case sensitivity in Snowflake, but this appears to
work.

This was also tested via CLI queries against our Snowflake
instance, and everything works as expected.

### CODE REVIEW

**Comment by tlento** [comment]
Current dependencies on/for this PR:
* main
  * **PR #629** <a href="https://app.graphite.dev/github/pr/dbt-labs/metricflow/629" target="_blank">Graphite</a> 
    * **PR #630** <a href="https://app.graphite.dev/github/pr/dbt-labs/metricflow/630" target="_blank">Graphite</a> 
      * **PR #631** <a href="https://app.graphite.dev/github/pr/dbt-labs/metricflow/631" target="_blank">Graphite</a>  👈

This comment was auto-generated by [Graphite](https://app.graphite.dev/github/pr/dbt-labs/metricflow/631?utm_source=stack-comment).

**Comment by tlento** [comment]
Note: engine tests passed on this version of the PR: https://github.com/dbt-labs/metricflow/actions/runs/5398257491

Snyk failures are expected due to updates to pyproject.toml. #632 is stacked on top of this and shows no issues.

File: metricflow/engine/metricflow_engine.py [hunk 1]
```diff
@@ -574,6 +574,12 @@ def get_dimension_values(  # noqa: D
         result_dataframe = query_result.result_df
         if result_dataframe is None:
             return []
+        if get_group_by_values not in result_dataframe.columns:
+            # Snowflake likes upper-casing things in result output, so we lower-case all names and
```
Comment (WilliamDee) [inline metricflow/engine/metricflow_engine.py:578]: 🤨 this seems annoying

**Review by WilliamDee — APPROVED** [review]
Dang, the casing stuff is really annoying. Aside from those concerns, lgtm!

**Pull request merged**